server/src/main/java/org/apache/iotdb/db/query/dataset/udf/UDTFFragmentDataSet.java

```diff
@@ -0,0 +1,122 @@
+ */
+
+package org.apache.iotdb.db.query.dataset.udf;
+
+import org.apache.iotdb.db.exception.query.QueryProcessException;
+import org.apache.iotdb.db.query.pool.DataSetFragmentExecutionPoolManager;
+import org.apache.iotdb.db.query.udf.core.layer.RawQueryInputLayer;
+import org.apache.iotdb.db.query.udf.core.reader.LayerPointReader;
+import org.apache.iotdb.tsfile.read.common.RowRecord;
+import org.apache.iotdb.tsfile.read.query.dataset.QueryDataSet;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+public class UDTFFragmentDataSet extends QueryDataSet {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(UDTFFragmentDataSet.class);
+
+  private static final int BLOCKING_QUEUE_CAPACITY = 2;
+  private static final DataSetFragmentExecutionPoolManager
+      DATA_SET_FRAGMENT_EXECUTION_POOL_MANAGER = DataSetFragmentExecutionPoolManager.getInstance();
+
+  private final QueryDataSet fragmentDataSet;
+  private final BlockingQueue<Object[]> productionBlockingQueue;
+
+  private RowRecord[] rowRecords = null;
+  private int rowRecordsLength = 0;
+  private int rowRecordsIndexConsumed = -1;
+
+  private boolean hasNextRowRecords = true;
+
+  public UDTFFragmentDataSet(RawQueryInputLayer rawQueryInputLayer, LayerPointReader[] transformers)
+      throws QueryProcessException, IOException {
+    fragmentDataSet = new UDTFAlignByTimeDataSet(rawQueryInputLayer, transformers);
+    productionBlockingQueue = new LinkedBlockingQueue<>(BLOCKING_QUEUE_CAPACITY);
+    submitTask();
+  }
+
+  @Override
+  public boolean hasNextWithoutConstraint() {
+    try {
+      return rowRecords != null || tryToTakeNextRowRecords();
+    } catch (InterruptedException e) {
+      return onThrowable(e);
+    }
+  }
+
+  private boolean tryToTakeNextRowRecords() throws InterruptedException {
+    if (!hasNextRowRecords) {
+      return false;
+    }
+
+    Object[] production = productionBlockingQueue.take();
+
+    Object rowRecordArrayOrThrowable = production[0];
+    if (rowRecordArrayOrThrowable instanceof Throwable) {
+      return onThrowable((Throwable) rowRecordArrayOrThrowable);
+    }
+
+    rowRecords = (RowRecord[]) production[0];
+    rowRecordsLength = (int) production[1];
+    rowRecordsIndexConsumed = -1;
+    hasNextRowRecords = (boolean) production[2];
+
+    if (rowRecordsLength == 0) {
+      rowRecords = null;
+      // assert !hasNextRowRecords;
+      return false;
+    }
+
+    if (hasNextRowRecords) {
+      submitTask();
+    }
+
+    return true;
+  }
+
+  @Override
+  public RowRecord nextWithoutConstraint() {
+    RowRecord rowRecord = rowRecords[++rowRecordsIndexConsumed];
+    if (rowRecordsIndexConsumed == rowRecordsLength - 1) {
+      rowRecords = null;
+    }
+    return rowRecord;
+  }
+
+  private void submitTask() {
+    if (productionBlockingQueue.remainingCapacity() > 0) {
```

Comment:
What's the consquence of running a query when `productionBlockingQueue.remainingCapacity() == 0`? It's more likely that the task will be dropped and the next time we call `tryToTakeNextRowRecords()`, it will be blocked in `productionBlockingQueue.take()` forever(`hasNextRoweRecords == true` and the previous call of `tryToTakeNextRowRecords()` is true). I think it's better to throw an exception when the queue is full.